controllers/capabilities/tenant_controller.go

```diff
@@ -98,6 +97,61 @@ func (r *TenantReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return ctrl.Result{}, err
 	}
 
+	// Tenant has been marked for deletion
+	if tenantR.GetDeletionTimestamp() != nil && controllerutil.ContainsFinalizer(tenantR, tenantFinalizer) {
+		existingTenant, err := controllerhelper.FetchTenant(tenantR, portaClient)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// delete tenantCR if tenant is present in 3scale
+		if existingTenant != nil {
+			// do not attempt to delete tenant that is already scheduled for deletion
+			if existingTenant.Signup.Account.State != scheduledForDeletionState {
+				err := controllerhelper.DeleteTenant(tenantR, portaClient)
+				if err != nil {
+					r.EventRecorder.Eventf(tenantR, corev1.EventTypeWarning, "Failed to delete tenant", "%v", err)
+					return ctrl.Result{}, err
+				}
+			} else {
+				reqLogger.Info("Removing tenant CR - tenant is already scheduled for deletion %v", existingTenant.Signup.Account.ID)
+			}
+		}
+
+		// add or remove finalizer, tenant CR will be deleted if tenant in 3scale does not exists, if tenant is deleted or if it's already marked for deletion
+		err = controllerhelper.ReconcileFinalizers(tenantR, r.Client, tenantFinalizer)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// Ignore deleted resources, this can happen when foregroundDeletion is enabled
+	// https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion
+	if tenantR.GetDeletionTimestamp() != nil {
+		return ctrl.Result{}, nil
+	}
+
+	changed := tenantR.SetDefaults()
```

Comment:
After all the finalizer logic?

Reply:
it is about step order:
- finalizer logic
- setting defaults
- reconciliate 3scale tenants

Setting the finalizer should be first thing to do, otherwise while the operator is doing "things", the user could remove the CR and the operator would not be able to handle the deletion.